PR: Add Refresh button to editors from Variable Explorer

jitseniesen opened the pull request:
## Description of Changes

* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)
* [x] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

This adds a Refresh button to the array editor, the dataframe editor, the collection editor and the object explorer. 

Clicking the Refresh button in the array editor refreshes the editor with the current value of the array variable:

https://github.com/spyder-ide/spyder/assets/7941918/a481ba3b-544b-4b93-8cc6-a6c43d4cd13b

The same for the dataframe editor:

https://github.com/spyder-ide/spyder/assets/7941918/ff8ba761-909b-407d-9fde-c45b03e59f70

The same for the collection editor:

https://github.com/spyder-ide/spyder/assets/7941918/15c51175-185e-454d-a0e0-5865196157e6

The same of the object explorer:

https://github.com/spyder-ide/spyder/assets/7941918/41e88831-63a9-4566-87c9-05832548b0a5

It is possible to refresh editors opened from the Variable Explorer or editors opened from other editors, so the user can for instance refresh an array nested inside a dictionary inside an object inside another object inside a list:

https://github.com/spyder-ide/spyder/assets/7941918/cbcf73d8-9488-48dc-8401-a164aa19b5d0

However, you can only refresh editors which ultimately originate from the Variable Explorer. For instance, you can open a collection editor from the IPython Console to display the contents of `sys.path`, but the Refresh button on that editor is disabled:

https://github.com/spyder-ide/spyder/assets/7941918/fe01b1bf-18bc-45e6-9a96-276c5e6b52d6

If the user changes a value in the editor and then refreshes, Spyder asks for confirmation:

https://github.com/spyder-ide/spyder/assets/7941918/ff0e2126-0ae2-4335-bca9-1640ae632ba1

If the new value can't be displayed in the array editor, for instance because the variable now contains an integer, then Spyder shows an error message and closes the array editor:

https://github.com/spyder-ide/spyder/assets/7941918/fc1c9947-be55-496e-a77c-5ab472f7ec82

The same happens if the variable no longer exists:

https://github.com/spyder-ide/spyder/assets/7941918/1d2b4d16-7ce0-46df-8a38-c818351e8f81

Note that in this case, the console displays an exception generated inside spyder-kernels. This is existing behaviour, the same happens in the current master if you open a variable in the Variable Explorer that no longer exists.

In the array editor, the UI changes depending on whether the array is a standard 2d array, a record array, a masked array or a 3d array. Refreshing the editor changes the UI if necessary:

https://github.com/spyder-ide/spyder/assets/7941918/7eeb33dc-d1dc-459e-9b70-f9aed79663cf

However, if the variable changes from an array to a dataframe, for instance, then the array editor will not transform into a dataframe editor. Instead, an informative message is displayed and the editor is closed, just as above where the array variable changed to an integer.

As part of this PR, I split the function which sets up the editors in two: one function for setting up the UI and one function for setting the value to be displayed. The first function is called only one, the second function is called when the editor is opened and when it is refreshed. I implemented the refresh functionality by passing another parameter, `data_function`, to the editors. Calling this function should return the current value of the variable; if `data_function` is `None` then the editor can't be refreshed.

The four editors (array, dataframe, collection and object) are all implemented differently, so the details of how to implement the refresh is also different in each case.

This PR removes the options `xlabels` and `ylabels` in the Array Editor. These can be used to change the column and row labels in the editor, but they are not currently used in Spyder. These options were added by Pierre in 2011. I could not find where they were used in the past, if they ever were, so I don't know their intended use. I don't know what to do when refreshing means that the size of the array changes, because then the number of labels also needs to change, so I chose to remove the options.


### Issue(s) Resolved

<!--- List the issue(s) below, in the form "Fixes #1234"; one per line --->

Fixes #6325 


### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

<!--- TYPE YOUR USER/NAME AFTER THE FOLLOWING: --->
I certify the above statement is true and correct: Jitse Niesen

<!--- Thanks for your help making Spyder better for everyone! --->

jitseniesen commented:
The current plan is to add a toolbar (as suggested in #21315) and then have this be a button in the toolbar. This matches the interface for the Variable Explorer.

pep8speaks commented:
Hello @jitseniesen! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`spyder/plugins/variableexplorer/widgets/arrayeditor.py`](https://github.com/spyder-ide/spyder/blob/988b6b1521e1e516275547bd7412c03375be7169/spyder/plugins/variableexplorer/widgets/arrayeditor.py):

> [Line 646:30](https://github.com/spyder-ide/spyder/blob/988b6b1521e1e516275547bd7412c03375be7169/spyder/plugins/variableexplorer/widgets/arrayeditor.py#L646): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 647:31](https://github.com/spyder-ide/spyder/blob/988b6b1521e1e516275547bd7412c03375be7169/spyder/plugins/variableexplorer/widgets/arrayeditor.py#L647): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'

* In the file [`spyder/plugins/variableexplorer/widgets/collectionsdelegate.py`](https://github.com/spyder-ide/spyder/blob/988b6b1521e1e516275547bd7412c03375be7169/spyder/plugins/variableexplorer/widgets/collectionsdelegate.py):

> [Line 97:80](https://github.com/spyder-ide/spyder/blob/988b6b1521e1e516275547bd7412c03375be7169/spyder/plugins/variableexplorer/widgets/collectionsdelegate.py#L97): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)







* In the file [`spyder/plugins/variableexplorer/widgets/objectexplorer/toggle_column_mixin.py`](https://github.com/spyder-ide/spyder/blob/988b6b1521e1e516275547bd7412c03375be7169/spyder/plugins/variableexplorer/widgets/objectexplorer/toggle_column_mixin.py):

> [Line 160:31](https://github.com/spyder-ide/spyder/blob/988b6b1521e1e516275547bd7412c03375be7169/spyder/plugins/variableexplorer/widgets/objectexplorer/toggle_column_mixin.py#L160): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ':'

* In the file [`spyder/plugins/variableexplorer/widgets/tests/test_arrayeditor.py`](https://github.com/spyder-ide/spyder/blob/988b6b1521e1e516275547bd7412c03375be7169/spyder/plugins/variableexplorer/widgets/tests/test_arrayeditor.py):

> [Line 245:5](https://github.com/spyder-ide/spyder/blob/988b6b1521e1e516275547bd7412c03375be7169/spyder/plugins/variableexplorer/widgets/tests/test_arrayeditor.py#L245): [E731](https://duckduckgo.com/?q=pep8%20E731) do not assign a lambda expression, use a def
> [Line 263:17](https://github.com/spyder-ide/spyder/blob/988b6b1521e1e516275547bd7412c03375be7169/spyder/plugins/variableexplorer/widgets/tests/test_arrayeditor.py#L263): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 265:5](https://github.com/spyder-ide/spyder/blob/988b6b1521e1e516275547bd7412c03375be7169/spyder/plugins/variableexplorer/widgets/tests/test_arrayeditor.py#L265): [E731](https://duckduckgo.com/?q=pep8%20E731) do not assign a lambda expression, use a def
> [Line 289:5](https://github.com/spyder-ide/spyder/blob/988b6b1521e1e516275547bd7412c03375be7169/spyder/plugins/variableexplorer/widgets/tests/test_arrayeditor.py#L289): [E731](https://duckduckgo.com/?q=pep8%20E731) do not assign a lambda expression, use a def

* In the file [`spyder/plugins/variableexplorer/widgets/tests/test_dataframeeditor.py`](https://github.com/spyder-ide/spyder/blob/988b6b1521e1e516275547bd7412c03375be7169/spyder/plugins/variableexplorer/widgets/tests/test_dataframeeditor.py):

> [Line 411:1](https://github.com/spyder-ide/spyder/blob/988b6b1521e1e516275547bd7412c03375be7169/spyder/plugins/variableexplorer/widgets/tests/test_dataframeeditor.py#L411): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

jitseniesen commented:
I extended this a fair bit. Now it is possible to refresh array editors, dataframe editors, collection editors (for list, set, dict and tuple), and object explorers (for any type that is not handled by another editor). It works for editors opened directly from the variable explorer and also for editors opened from collection and object editors.

The PR conflicts with PR #20546 and PR #21317, which add a toolbar to the dataframe editor and array editor respectively. The plan is to merge the two PRs that add a toolbar first, and then change this PR so that the Refresh button is added to the toolbar.

ccordoba12 commented:
> I extended this a fair bit. Now it is possible to refresh array editors, dataframe editors, collection editors (for list, set, dict and tuple), and object explorers (for any type that is not handled by another editor).

Great work @jitseniesen! That'd be really nice to have.

> The PR conflicts with PR https://github.com/spyder-ide/spyder/pull/20546 and PR https://github.com/spyder-ide/spyder/pull/21317

Ok, I'll try to review soon those PRs so you can finish this one.

jitseniesen commented:
I rebased the PR in order to pick up the addition of the toolbar to the array and dataframe editors (thanks Carlos for the speedy review) and I updated the screenshots in the top comment. This is now ready for review.

jitseniesen commented:
I rebased to pick up the CI automatic test fix in PR #21517.

dalthviz reviewed (commented):
Thanks @jitseniesen ! I checked locally this and seems like things are working as expected :+1:

Left a single comment regarding the `safe_disconnect` function definition. Seems like taht function could be worthy to be moved to some `utils` module for reusage (not only for this PR but seems like there are other places inside the source code from a quick search which could benefit from using it)

Other than that this LGTM :+1:

jitseniesen commented on spyder/plugins/variableexplorer/widgets/objectexplorer/objectexplorer.py:
```diff
@@ -42,6 +43,15 @@
+        signal.disconnect()
+    except TypeError:
+        # Raised when no slots are connected to the signal
+        pass
```
That's a good idea, I added a commit implementing it.

GitKia1392 commented:
Can you provide the Videos in mp3 format, MIME format is unstable and unreadable on Windows 11 or Typically Every windows + xp

Appreciated
Kiamehr Esk

dalthviz commented:
Regarding the tests not passing, maybe a rebase on top of the latest `master` could help @jitseniesen ? Checking the commits history seems like in a recent commit some changes were done to the CI related with test env activation and `spyder-kernels`: https://github.com/spyder-ide/spyder/commit/9c07bc8b8320abbf929acadf0334592bcec578e1

dalthviz reviewed (approved):
Thanks @jitseniesen !

jitseniesen commented:
> maybe a rebase on top of the latest `master` could help

@dalthviz Thanks a lot for this suggestion. I had somehow convinced myself that rebasing could not possibly make a difference, so I kept looking for other explanations.

ccordoba12 commented on spyder/plugins/variableexplorer/widgets/tests/test_dataframeeditor.py:
```diff
@@ -427,6 +434,90 @@ def test_dataframemodel_with_format_percent_d_and_nan():
+
+def test_dataframeeditor_refresh():
+    """
+    Test that after pressing the refresh button, the value of the Array Editor
+    is replaced by the return value of the data_function.
```
```suggestion
    Test that after pressing the refresh button, the value of the editor is
    replaced by the return value of the data_function.
```

ccordoba12 commented on spyder/utils/qthelpers.py:
```diff
@@ -886,5 +886,14 @@ def handle_applicationStateChanged(state):
 
 
+def safe_disconnect(signal):
+    """Disconnect a QtSignal, ignoring TypeError"""
```
```suggestion
    """Disconnect a Qt signal, ignoring TypeError."""
```

ccordoba12 commented on spyder/plugins/variableexplorer/widgets/collectionsdelegate.py:
```diff
@@ -414,8 +462,11 @@ def updateEditorGeometry(self, editor, option, index):
+    def __init__(self, parent=None, namespacebrowser=None,
+                 data_function: Optional[Callable[[], Any]] = None):
+        CollectionsDelegate.__init__(
+            self, parent, namespacebrowser, data_function)
```
```suggestion
            self, parent, namespacebrowser, data_function
        )
```

ccordoba12 commented on spyder/plugins/variableexplorer/widgets/objectexplorer/toggle_column_mixin.py:
```diff
@@ -155,12 +156,15 @@ class ToggleColumnTreeView(QTreeView, ToggleColumnMixIn):
         from spyder.plugins.variableexplorer.widgets.collectionsdelegate \
             import ToggleColumnDelegate
-        self.delegate = ToggleColumnDelegate(self, namespacebrowser)
+        self.delegate = ToggleColumnDelegate(
+            self, namespacebrowser, data_function)
```
```suggestion
        self.delegate = ToggleColumnDelegate(
            self, namespacebrowser, data_function
        )
```

ccordoba12 commented on spyder/widgets/collectionseditor.py:
```diff
@@ -1332,7 +1334,8 @@ def __init__(self, parent, data, namespacebrowser=None, readonly=False,
         self.setModel(self.source_model)
-        self.delegate = CollectionsDelegate(self, namespacebrowser)
+        self.delegate = CollectionsDelegate(
+            self, namespacebrowser, data_function)
```
```suggestion
            self, namespacebrowser, data_function
        )
```

ccordoba12 commented on spyder/widgets/collectionseditor.py:
```diff
@@ -1444,15 +1447,18 @@ def set_filter(self, dictfilter=None):
         else:
             self.editor = CollectionsEditorTableView(
-                self, data, namespacebrowser, readonly, title)
+                self, data, namespacebrowser, data_function, readonly, title)
```
```suggestion
                self, data, namespacebrowser, data_function, readonly, title
            )
```

ccordoba12 commented on spyder/widgets/collectionseditor.py:
```diff
@@ -1461,8 +1467,18 @@ def __init__(self, parent, data, namespacebrowser=None, readonly=False,
+            text=_('Refresh'),
+            icon=ima.icon('refresh'),
+            tip=_('Refresh editor with current value of variable in console'),
+            triggered=lambda: self.sig_refresh_requested.emit())
```
```suggestion
            triggered=lambda: self.sig_refresh_requested.emit()
        )
```

ccordoba12 commented on spyder/widgets/collectionseditor.py:
```diff
@@ -1521,8 +1539,9 @@ def setup(self, data, title='', readonly=False, remote=False,
-            self, self.data_copy, self.namespacebrowser, title=title,
-            readonly=readonly, remote=remote)
+            self, self.data_copy, self.namespacebrowser, self.data_function,
+            title=title, readonly=readonly, remote=remote)
```
```suggestion
            title=title, readonly=readonly, remote=remote
        )
```

ccordoba12 commented on spyder/widgets/collectionseditor.py:
```diff
@@ -1573,6 +1592,45 @@ def get_value(self):
+        try:
+            new_value = self.data_function()
+        except (IndexError, KeyError):
+            QMessageBox.critical(self, _('Collection editor'),
+                                 _('The variable no longer exists.'))
```
```suggestion
            QMessageBox.critical(
                self,
                _('Collection editor'),
                _('The variable no longer exists.')
            )
```

ccordoba12 commented on spyder/widgets/collectionseditor.py:
```diff
@@ -1573,6 +1592,45 @@ def get_value(self):
+        message = _('Refreshing the editor will overwrite the changes that '
+                    'you made. Do you want to proceed?')
+        result = QMessageBox.question(
+            self, _('Refresh collection editor?'), message)
```
```suggestion
            self,
            _('Refresh collections editor?'),
            message
        )
```

ccordoba12 commented on spyder/widgets/collectionseditor.py:
```diff
@@ -1595,6 +1653,36 @@ def set_value(self, index, value):
             name = source_index.model().keys[source_index.row()]
             self.parent().new_value(name, value)
 
+    def make_data_function(self, index: QModelIndex
+                           ) -> Optional[Callable[[], Any]]:
```
```suggestion
    def make_data_function(
        self,
        index: QModelIndex
    ) -> Optional[Callable[[], Any]]:
```

ccordoba12 commented on spyder/widgets/tests/test_collectioneditor.py:
```diff
@@ -265,6 +265,25 @@ def test_filter_rows(qtbot):
 
+def test_remote_make_data_function():
+    """
+    Test that the function returned by make_data_function() ...
```
```suggestion
    Test that the function returned by make_data_function() is the expected
    one.
```

I'm not sure if this is really what this docstring should say. If not, please edit my suggestion before applying it.

ccordoba12 commented on spyder/widgets/tests/test_collectioneditor.py:
```diff
@@ -265,6 +265,25 @@ def test_filter_rows(qtbot):
+                       'numpy_type': 'Unknown'}}
+    mock_shellwidget = Mock()
+    editor = RemoteCollectionsEditorTableView(
+        None, variables, mock_shellwidget)
```
```suggestion
        None, variables, mock_shellwidget
    )
```

ccordoba12 commented on spyder/widgets/tests/test_collectioneditor.py:
```diff
@@ -478,6 +497,91 @@ def test_rename_and_duplicate_item_in_collection_editor():
+
+def test_collectioneditor_refresh():
+    """
+    Test that after pressing the refresh button, the value of the Array Editor
+    is replaced by the return value of the data_function.
```
```suggestion
    Test that after pressing the refresh button, the value of the editor is
    replaced by the return value of the data_function.
```

ccordoba12 commented on spyder/widgets/tests/test_collectioneditor.py:
```diff
@@ -478,6 +497,91 @@ def test_rename_and_duplicate_item_in_collection_editor():
+@pytest.mark.parametrize('result', [QMessageBox.Yes, QMessageBox.No])
+def test_collectioneditor_refresh_after_edit(result):
+    """
+    Test that after changing a value in the collection editor, refreshing the
```
```suggestion
    Test that after changing a value in the collections editor, refreshing the
```

ccordoba12 commented on spyder/widgets/tests/test_collectioneditor.py:
```diff
@@ -478,6 +497,91 @@ def test_rename_and_duplicate_item_in_collection_editor():
+    """
+    Test that if the variable is deleted and then the editor is refreshed
+    (resulting in data_function raising a KeyError), a critical dialog box
+    is displayed and that the array editor is closed.
```
```suggestion
    is displayed and that the editor is closed.
```

ccordoba12 commented on spyder/widgets/tests/test_collectioneditor.py:
```diff
@@ -478,6 +497,91 @@ def test_rename_and_duplicate_item_in_collection_editor():
+    lst = [1, 2, 3, 4]
+    editor = CollectionsEditor(None, data_function=datafunc)
+    editor.setup(lst)
+    with patch('spyder.plugins.variableexplorer.widgets.arrayeditor'
+               '.QMessageBox.critical') as mock_critical, \
```
```suggestion
    with patch('spyder.widgets.collectionseditor.QMessageBox'
               '.critical') as mock_critical, \
```

ccordoba12 commented on spyder/plugins/variableexplorer/widgets/objectexplorer/objectexplorer.py:
```diff
@@ -197,6 +254,14 @@ def _setup_menu(self, show_callable_attributes=False,
+            self, icon=ima.icon('refresh'),
+            tip=_('Refresh editor with current value of variable in console'),
+            triggered=self.refresh_editor)
+        self.refresh_button.setEnabled(self.data_function is not None)
```
```suggestion
        self.refresh_button.setEnabled(self.data_function is not None)
        self.refresh_button.setStyleSheet(str(PANES_TOOLBAR_STYLESHEET))
```

ccordoba12 commented:
Note: This also needs a rebase to get the fixes I did in PR #21553.

ccordoba12 reviewed (approved):
Looks good to me now. Thanks @jitseniesen for this terrific improvement!

rpinto73 commented:
Thank you so much to @jitseniesen and everyone else who worked hard to make this feature a reality!!
